crates/iota-grpc-api/src/event_service.rs

```diff
@@ -0,0 +1,231 @@
+            build_or_filter(filters)
+        }
+        Some(Filter::All(_)) => Ok(EventFilter::All(vec![])),
+        Some(Filter::Sender(f)) => Ok(EventFilter::Sender(parse_address(&f.sender)?)),
+        Some(Filter::Transaction(f)) => {
+            Ok(EventFilter::Transaction(parse_tx_digest(&f.tx_digest)?))
+        }
+        Some(Filter::MoveModule(f)) => Ok(EventFilter::MoveModule {
+            package: parse_object_id(&f.package_id)?,
+            module: parse_identifier(&f.module, "module name")?,
+        }),
+        Some(Filter::TimeRange(f)) => Ok(EventFilter::TimeRange {
+            start_time: f.start_time,
+            end_time: f.end_time,
+        }),
+        None => Ok(EventFilter::All(vec![])),
+    }
+}
+
+// Helper functions to reduce repetition and improve error messages
+fn parse_object_id(hex_str: &str) -> Result<ObjectID, Status> {
+    ObjectID::from_hex_literal(hex_str)
+        .map_err(|e| Status::invalid_argument(format!("Invalid object ID '{hex_str}': {e}")))
+}
+
+fn parse_identifier(id_str: &str, field_name: &str) -> Result<Identifier, Status> {
+    Identifier::from_str(id_str)
+        .map_err(|e| Status::invalid_argument(format!("Invalid {field_name} '{id_str}': {e}")))
+}
+
+fn parse_address(addr_str: &str) -> Result<IotaAddress, Status> {
+    IotaAddress::from_str(addr_str)
+        .map_err(|e| Status::invalid_argument(format!("Invalid address '{addr_str}': {e}")))
+}
+
+fn parse_tx_digest(digest_str: &str) -> Result<TransactionDigest, Status> {
+    TransactionDigest::from_str(digest_str).map_err(|e| {
+        Status::invalid_argument(format!("Invalid transaction digest '{digest_str}': {e}"))
+    })
+}
+
+fn parse_filter_list(filters: &[crate::events::EventFilter]) -> Result<Vec<EventFilter>, Status> {
+    filters.iter().map(create_event_filter).collect()
+}
+
+/// Generic function to build chained filters to ease of implementing other
+/// logics in the future
+fn build_chained_filter<F>(
+    mut filters: Vec<EventFilter>,
+    filter_type: &str,
+    constructor: F,
+) -> Result<EventFilter, Status>
+where
+    F: Fn(Box<EventFilter>, Box<EventFilter>) -> EventFilter,
+{
+    if filters.is_empty() {
+        return Err(Status::invalid_argument(format!(
+            "{filter_type} filter cannot be empty",
+        )));
+    }
+
+    let mut result = filters.pop().unwrap();
+    while let Some(filter) = filters.pop() {
+        result = constructor(Box::new(filter), Box::new(result));
+    }
+    Ok(result)
+}
+
+/// Build AND filter by chaining EventFilter::And
+fn build_and_filter(filters: Vec<EventFilter>) -> Result<EventFilter, Status> {
+    build_chained_filter(filters, "AND", EventFilter::And)
+}
+
+/// Build OR filter by chaining EventFilter::Or
+fn build_or_filter(filters: Vec<EventFilter>) -> Result<EventFilter, Status> {
+    build_chained_filter(filters, "OR", EventFilter::Or)
+}
+
+// Convert IotaEvent to protobuf Event
+impl TryFrom<&IotaEvent> for Event {
```

Comment:
Should it be unfallible?
```suggestion
impl From<&IotaEvent> for Event {
```

Reply:
Agreed. I also refactored all the Event protobuf and also make it unfallible in 8dc4a7a61d433c863eaa023bdf32d1329fd25262